rfcs/0000-custom-health-checks/README.md

```diff
@@ -0,0 +1,330 @@
+#### Provide an evaluator for `CEL` expressions for users
+
+We will provide a CEL environment that can be used by the user to evaluate `CEL`
+expressions. Users will use it to test their expressions before applying them to
+their `Kustomization` object.
+
+```shell
+$ flux eval --api-version cert-manager.io/v1 --kind Certificate --in-progress "status.conditions.filter(e, e.type == 'Issuing').all(e, e.observedGeneration == metadata.generation && e.status == 'True')" --failed "status.conditions.filter(e, e.type == 'Ready').all(e, e.observedGeneration == metadata.generation && e.status == 'False')" --current "status.conditions.filter(e, e.type == 'Ready').all(e, e.observedGeneration == metadata.generation && e.status == 'True')" --file ./custom_resource.yaml
+```
+
+### User Stories
+
+#### Configure custom health checks for a custom resource
+
+> As a user of Flux, I want to be able to specify custom health checks for my
+> custom resources, so that I can have more control over the status of my
+> resources.
+
+#### Enable health checks support in Flux for non-standard resources
+
+> As a user of Flux, I want to be able to use the health check feature for
+> non-standard resources, so that I can have more control over the status of my
+> resources.
+
+### Alternatives
+
+We need an expression language that is flexible enough to cover all possible use
+cases, without having to change `Flux` source code for each new use case.
+
+On alternative that have been considered is to use `cuelang` instead of `CEL`.
+`cuelang` is a more powerful expression language, but it is also more complex and 
+requires more work to integrate with `Flux`. it also does not have any support in
+`Kubernetes` yet while `CEL` is already used in `Kubernetes` and libraries are
+available to use it.
+
+## Design Details
+
+### Introduce a new field `CustomHealthChecksExprs` in the `Kustomization` CRD
+
+The `api/v1/kustomization_types.go` file will be updated to add the `CustomHealthChecksExprs`
+field to the `KustomizationSpec` struct.
+
+```go
+type KustomizationSpec struct {
+...
+	// A list of resources to be included in the health assessment.
+	// +optional
+	HealthChecks []meta.NamespacedObjectKindReference `json:"healthChecks,omitempty"`
+
+	// A list of custom health checks expressed as CEL expressions.
+	// The CEL expression must evaluate to a boolean value.
+	// +optional
+	CustomHealthChecksExprs []CustomHealthCheckExprs `json:"customHealthChecksExprs,omitempty"`
+...
+}
+
+// CustomHealthCheckExprs defines the CEL expressions for custom health checks.
+// The CEL expressions must evaluate to a boolean value. The expressions are used
+// to determine the status of the custom resource.
+type CustomHealthCheckExprs struct {
+	// apiVersion of the custom health check.
+	// +required
+	APIVersion string `json:"apiVersion"`
+	// Kind of the custom health check.
+	// +required
+	Kind string `json:"kind"`
+	// InProgress is the CEL expression that verifies that the status
```

Comment:
I'm thinking that it would be nice to here have a `Name` field, to allow definitng a healthCheck applicable to a specific resource, with a finer grained granularity than the CRD.

The kind of scenario where this would be relevant would be situations where a Kustomization would produce multiple resources of the same kind but the right healthcheck to do would be different among those resources.

An optional `Name` field would be interpreted as: the healthCheck to use for a resource of Kind Y and Name X, would be the item in CustomHealthCheckExprs that specifies Kind Y and Name X if any, or if there is none an item for Kind Y and Name unspecified.

Reply:
I will gladly admit not having any exact use case in mind.

The idea expressed [here](https://github.com/fluxcd/flux2/pull/4528/files#r1463040255) might be one, though, but whether or not it's relevant is certainly debatable.  The intersection of "I want this Deployment to be healhcheck'd in a specific way" and "my Kustomization produces multiple Deploments" is perhaps empty.

From a user standpoint, it simply feels like "if this is going to be extensible, why not make it fully flexible".  But from an implementation and code maintenance standpoint, I would understand that this idea might  be perceived as over-engineering.

I just wanted to share this to see how relevant and implementable the idea might be.

Reply:
sorry for the slow feedback

yes @souleb there would be workarounds

and this is overall a very useful evolution to what we have today which works pretty well for many cases already, so the wise path is of course, to keep it simple